File: Palestra di Barbara Liskov (OOP)/Battle Arena/Team11/BattleArena.py
from enum import Enum
import random

class Type(Enum):
    fire = 1
    water = 2
    grass = 3

efficacy = {
    Type.fire: {Type.grass: 2.0, Type.water: 0.5},
    Type.grass: {Type.water: 2.0, Type.fire: 0.5},
    Type.water: {Type.fire: 2.0, Type.grass: 0.5},
}

class Pokemon:
    name = ""
    type = ""
    hp = 0
    moves = []
    stats = {}
    def _init_(self,name,type,hp,moves,stats):
        self.name = name 
        self.type = type
        self.hp = hp
        self.moves = moves
        self.stats = stats
    
    def get_random_move(self):
        return random.choice(self.moves) if self.moves else None
    
    def attack(self,targert_pokemon, move):
        print(self.name," uses ", move.name, " on ", targert_pokemon.name )
        multiplier = efficacy.get(move.type, {}).get(targert_pokemon.type)
        damage = move.damage * multiplier
        targert_pokemon.defend(damage)

    def defend(self,damage):
        self.hp = max(self.hp-damage,0)
        print(self.name, " takes ", damage , " damage. ", self.hp, " HP left.")

class Move:
    name = ""
    type = ""
    damage = 0
    pp = 0
    def _init_ (self,name,type,damage,pp):
        self.name = name
        self.type = type
        self.damage = damage
        self.pp = pp

class Battle:
    def startBattle(self, pokemon1,pokemon2):
        while(pokemon1.hp > 0 and pokemon2.hp>0):
            pokemon1.attack(pokemon2,pokemon1.get_random_move() )
            if(pokemon2.hp >0):
                pokemon2.attack(pokemon1,pokemon2.get_random_move() )
        if(pokemon2.hp > pokemon1.hp):
            print(pokemon2.name, " win")
        else:
            print(pokemon1.name, " win")

def main():
    ember = Move("ember",Type.fire,30,15)
    hydropump = Move("hydropump",Type.water,15,10)
    bulletseed = Move("bulletseed",Type.grass,5,5)
    charmander = Pokemon("charmander",Type.fire,60,[ember],{'Attack':123, 'Speed':23, 'Defence':40})
    bulbasaur = Pokemon("bulbasaur",Type.grass,100,[bulletseed],{'Attack':12, 'Speed':3, 'Defence':5})

    battle = Battle()
    battle.startBattle(charmander,bulbasaur)

main()

File: Palestra di Barbara Liskov (OOP)/Battle Arena/BattleArena.java
class Pokemon {
    private String name;
    private String type;
    private int hp;
    private List<String> moves;
    private Map<String, Integer> stats;

    public Pokemon(String name, String type, int hp, List<String> moves, Map<String, Integer> stats) {
        this.name = name;
        this.type = type;
        this.hp = hp;
        this.moves = moves;
        this.stats = stats;
    }

    public void attack(Pokemon target, String move) {
        System.out.println(this.name + " uses " + move + " on " + target.getName() + "!");
        int damage = calculateDamage(move, target.getType());
        target.defend(damage);
    }

    private int calculateDamage(String move, String targetType) {
        // Qui puoi implementare la logica per calcolare il danno basato sulla mossa e sul tipo del bersaglio
        return 10; // Esempio di valore fisso per semplicità
    }

    public void defend(int damage) {
        this.hp -= damage;
        System.out.println(this.name + " takes " + damage + " damage. " + this.hp + " HP left.");
    }

    public String getName() {
        return name;
    }

    public String getType() {
        return type;
    }

    public int getHp() {
        return hp;
    }
}

class Battle {
    public void startBattle(Pokemon pokemon1, Pokemon pokemon2) {
        while (pokemon1.getHp() > 0 && pokemon2.getHp() > 0) {
            pokemon1.attack(pokemon2, "SELECTED_MOVE"); // Sostituisci SELECTED_MOVE con la mossa scelta
            if (pokemon2.getHp() > 0) {
                pokemon2.attack(pokemon1, "SELECTED_MOVE"); // Sostituisci SELECTED_MOVE con la mossa scelta
            }
        }
        if (pokemon1.getHp() > 0) {
            System.out.println(pokemon1.getName() + " wins!");
        } else {
            System.out.println(pokemon2.getName() + " wins!");
        }
    }
}